pkg/storage/internal/repo/certificates.go

```diff
@@ -0,0 +1,124 @@
+package repo
+
+import (
+	"fmt"
+
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/utils/to"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/storage/internal/database/models"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/storage/internal/database/scopes"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/storage/internal/paging"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/wdk"
+	"gorm.io/gorm"
+)
+
+type Certificates struct {
+	db *gorm.DB
+}
+
+type ListCertificatesActionParams struct {
+	SerialNumber       *wdk.Base64String
+	Subject            *wdk.PubKeyHex
+	RevocationOutpoint *wdk.OutpointString
+	Signature          *wdk.HexString
+	Certifiers         []wdk.PubKeyHex
+	Types              []wdk.Base64String
+	Limit              wdk.PositiveIntegerDefault10Max10000
+	Offset             wdk.PositiveIntegerOrZero
+}
+
+func NewCertificates(db *gorm.DB) *Certificates {
+	return &Certificates{db: db}
+}
+
+func (c *Certificates) CreateCertificate(certificate *models.Certificate) (uint, error) {
+	err := c.db.Create(certificate).Error
+	if err != nil {
+		return 0, fmt.Errorf("failed to create certificate model: %w", err)
+	}
+	return certificate.ID, nil
+}
+
+func (c *Certificates) DeleteCertificate(userID int, args wdk.RelinquishCertificateArgs) error {
+	tx := c.db.Delete(&models.Certificate{}, "type = ? AND serial_number = ? AND certifier = ? AND user_id = ?", args.Type, args.SerialNumber, args.Certifier, userID)
+	if tx.RowsAffected == 0 {
+		return fmt.Errorf("failed to delete certificate model: certificate not found")
+	}
+	if tx.Error != nil {
+		return fmt.Errorf("failed to delete certificate model: %w", tx.Error)
+	}
+
+	return nil
+}
+
+func (c *Certificates) ListAndCountCertificates(userID int, opts ListCertificatesActionParams) ([]*models.Certificate, int64, error) {
+	var certificates []*models.Certificate
+	var totalRows int64
```

Comment:
maybe use the named response params instead
```suggestion
func (c *Certificates) ListAndCountCertificates(userID int, opts ListCertificatesActionParams) (certificates []*models.Certificate, totalRows int64, err error) {
```